### PR TITLE
Tempo: Support TraceQL metrics queries

### PR DESCRIPTION
**Pull request by adrapereira**
**What is this feature?**

This PR adds support for TraceQL metrics queries in the TraceQL tab of the Tempo datasource by detecting when the query contains a metric function and calling the appropriate Tempo endpoint. The response is formatted as a time series compatible data frame, with preferred visualisation type as `graph`. 

Also has a small refactor of the TraceQL section of the `query()` function of the Tempo datasource, to make it more readable since it can execute a few different types of queries based on the content of the editor. 

**Why do we need this feature?**

Supports Tempo's newest feature and allows users to use the metrics endpoint in dashboards or explore the data. 

**Who is this feature for?**

Users of the Tempo datasource.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/datasource/tempo/datasource.ts
Comment (fabrizio-grafana) [reply public/app/plugins/datasource/tempo/datasource.ts:485]: > Detect when the query contains a metrics function

Maybe we can add this as a code comment. `isTraceIdQuery` already has one, so it could make sense also for `isTraceQlMetricsQuery`.

**Review by fabrizio-grafana — COMMENTED** [review]
The approach is GTM, but let's maybe wait for some other reviews.
Once the approach is approved by most of us / everyone, would it be possible to add a few unit tests?

File: public/app/plugins/datasource/tempo/resultTransformer.ts
Comment (joey-grafana) [inline public/app/plugins/datasource/tempo/resultTransformer.ts:629]: Should we be more defensive in how we map response data? Something we should discuss, but not strictly necessary to be done in this PR. We expect data.series / samples to exist but honestly I think we should be more careful and support the edge case that perhaps the data has not returned correctly in case of a malformed response etc that is not caught as an error/warning somewhere already up the response chain.

File: public/app/plugins/datasource/tempo/resultTransformer.ts
Comment (joey-grafana) [inline public/app/plugins/datasource/tempo/resultTransformer.ts:640]: nit: clearer to do the mapping here rather than in const above

**Review by joey-grafana — COMMENTED** [review]
Few comments/suggestions but main one is for tests with dummy data :) 

If I add this to a dashboard, the table is selected by default but it should probably be the time series panel.

Also it seems that rate is not matching in the query regex
Screenshot 2024-02-07 at 14 51 42

We need to e.g. add `{status=error}` in front - Is this because we actually need something to get metrics for and that is why we first need the `{status=error}` part?
Screenshot 2024-02-07 at 14 53 35

**Comment by joey-grafana** [comment]
Forgot to mention, we will probably want to add more kinds of visualisation in the future (perhaps near future). How do we see allowing the user to select for example and scatter plot or bar chart instead of a time series? Is it going to just be a case of transforming the data differently or will we more need a switch in the UI (perhaps in the query options).

**Comment by joey-grafana** [comment]
> @joey-grafana From what I can tell from the TraceQL metrics proposal `rate() by (...)` is not a valid query, you would have to change it to be `{} | rate() by (...)` to make it a valid query.

Thanks makes sense. Thanks for clarifying.

Two other things that came up:
- Seems only `rate` and `count_over_time` return results, the others return in error.
- Should this be behind a feature toggle?

**Comment by joey-grafana** [comment]
> They're the only functions implemented in Tempo for now, we'll have to wait for the other functions to be ready. Perhaps this is something we can show differently with syntax highlighting and/or grammar.

If they can't be used now then I'd imagine it's more of a convenience to add them to the list in this PR. I would not be in favor of treating them differently with syntax highlighting for that reason. As a matter of interest, when are they planned for and where did you get the list to begin with?

> Good point, I didn't consider alternate visualisations options, perhaps this is something we can explore in the next few weeks?

Will add our options to the design doc and we can all review.

**Review by joey-grafana — COMMENTED** [review]
FYI these may be issues, just want to make sure you've seen this and it's actually an issue with the graph and not the PR, otherwise ready to accept as PR LGTM!

Seeing hover tooltip but not hovering over any dot
Screenshot 2024-02-13 at 09 20 15

Hovering over blue dot but seeing app, which is green
Screenshot 2024-02-13 at 09 20 32

**Review by joey-grafana — APPROVED** [review]

**Pull request merged**
